### PR TITLE
Expand dark_current module for use with 1D data

### PR DESCRIPTION
**Pull request by hoogerheide**
The dark_current module is used for subtracting slit1-dependent dark counts in the candor loader, or as a standalone module (but for a single dataset).

This update expands the dark_current module to handle refldata[] lists from either MAGIK or CANDOR (not tested on other instruments), and allows it to be used in a workflow immediately after the loader (rather than requiring it to be incorporated into the loader).

It has been tested both on Candor data (only in the loader), and on Magik data as part of the workflow.

Also, because of its more general nature now, it was moved from candor_steps to steps.

The first commit in this PR creates a separate module for 1D data, subtract_dark_counts, that is replaced by the updated dark_current in the last commit.

Not done: Uncertainty propagation may still be important, or perhaps recalculating the uncertainties based on the dark-current-subtracted values. Also there's a TODO for datatype hierarchy that I am not sure what it means.

### CODE REVIEW

**Comment by hoogerheide** [comment]
An additional note on the dark current subtraction. Currently this is implemented with two parameters, a slope and an intercept of the dark current as a function of slit 1 opening. The intercept is the "true" dark current, while the slope represents a type of background arising from beam impinging on slit 2. In practice this is calibrated by closing slit 2 and then measuring the dark current as a function of slit 1 opening. This was found to be appropriate for Candor, where the post-sample collimation is essentially fixed, at the high Q regions where this all matters, by the 8 mm "slit 4" aperature. (Though, note that the Candor dark current has in fact been detector angle dependent, though increased shielding may prove to mitigate this somewhat.)

For MAGIK the dynamics are a little different. Many scans have slit 4 opening as the detector angle increases, so the  solid angle seen by the detector is changing, and the sensitivity to slit 1 opening increases with slit 4 opening. In this case the dark current is actually something like dark_current = (A * s4 + B) * s1 + C. If s4 opens proportionately to s1, this is quadratic in s1, and in fact this is what is observed on MAGIK; the dark current vs s1 slope is correlated with the value of s4.

image
image

The issue for the dark_current module would be back compatibility. My suggestion is to change the dark_current module to use as input a list of polynomial coefficients describing the dark current background field as a function of s1 opening. This allows back compatibility with the Candor loader because we can construct the linear polynomial in the candor loader using the same parameters as before. I will update this PR soon with a commit implementing this, but in the meantime any comments on this would be appreciated.

**Comment by bmaranville** [comment]
My (unrealistic?) expectation is that we shouldn't worry about breaking changes to CANDOR reduction until the instrument is in the user program.  If an instrument scientist needs to go back and do some gymnastics to reduce old data that's ok if we need to change the functions going forward.

**Comment by hoogerheide** [comment]
@bmaranville Okay, sounds good. Will update with a new commit. Plan is to allow a list input to dark_current that takes polynomial coefficients, as well as a list of lists input that represents the covariance matrix that we can use for error propagation.

**Comment by hoogerheide** [comment]
I think this is ready to go pending any further review.

**Comment by bmaranville** [comment]
@hoogerheide will this break existing reductions that use the earlier form of dark current entries in the loader?  Also, how often do you think people will need to use the polynomial form when doing reduction?  Is it an edge case, or will it become a standard protocol, in your opinion?

**Comment by hoogerheide** [comment]
tl;dr: 1. loader inputs haven't changed. 2. this is our chance to determinethe input structure for dark_current. Minimum quadratic in s1 is required; polynomial is most flexible and allows straightforward specification of the covariance matrix for errors.

No, the loader inputs (dc_slope, dc_rate) haven't changed. They're just massaged into the polynomial inputs that the dark_current module now accepts. The output might be slightly different because I no longer allow negative dark current or negative counts due to dark current subtraction.

As for whether or not this will be used: I'm not sure how important it is to do a dark current subtraction anyway, since most interpolation background subtraction schemes will take care of this. Certain schemes like the background field subtraction will make use of it because it requires dark count subtracted data to work.

As for the polynomial form: in my opinion, as soon as you have the dc_slope parameter you've moved away from the idea of "dark current", which really should be more or less independent of the state of the instrument, and toward a polynomial form. Since we already know that in the constant-footprint mode of operation on MAGIK dc_slope isn't enough, we need at minimum a quadratic. And in that case we might as well just allow a generic polynomial (though there's something to be said for using the form dark_current = (A * s4 + B) * s1 + C instead).

On CANDOR whether or not the polynomial form is required will depend on how much the slit1-dependence of the dark count rate has improved with the new shielding configuration. This remains to be measured.

The polynomial has the advantage that it requires only two (list) inputs to do the subtraction and the error propagation. In principle, if you had three inputs: quadratic coefficient, linear coefficient, and constant term, you would still need another input with the covariance matrix. Specifying this would get confusing and tricky, I think.

**Review by bmaranville — APPROVED** [review]
